astropy/io/fits/src/compressionmodule.c

```diff
@@ -233,159 +235,136 @@ int compress_type_from_string(char* zcmptype) {
-        *val = (int) PyLong_AsLong(keyval);
-        retval = 0;
-    }
-    else {
-        PyErr_Clear();
+        long tmp = PyLong_AsLong(keyval);
+        Py_DECREF(keyval);
+        if (PyErr_Occurred()) {
+            return -1;
+        }
+        *val = tmp;
+        return 0;
+    } else {
         *val = def;
-        retval = 1;
+        return PyErr_Occurred() ? -1 : 1;
     }
-
-    Py_DECREF(keystr);
-    Py_XDECREF(keyval);
-    return retval;
 }
 
 
-int get_header_long(PyObject* header, char* keyword, long* val, long def) {
-    PyObject* keystr;
-    PyObject* keyval;
-    int retval;
-
-    keystr = PyUnicode_FromString(keyword);
-    keyval = PyObject_GetItem(header, keystr);
-
-    if (keyval != NULL) {
-        *val = PyLong_AsLong(keyval);
-        retval = 0;
-    }
-    else {
-        PyErr_Clear();
-        *val = def;
-        retval = 1;
+int get_header_int(PyObject* header, char* keyword, int* val, int def) {
+    long tmp;
+    int ret = get_header_long(header, keyword, &tmp, def);
+    if (ret == 0) {
+        if (tmp > INT_MIN && tmp < INT_MAX) {
+            *val = (int) tmp;
+        } else {
+            PyErr_Format(PyExc_OverflowError, "Cannot convert %ld to C 'int'", tmp);
+            ret = -1;
+        }
     }
-
-    Py_DECREF(keystr);
-    Py_XDECREF(keyval);
-    return retval;
+    return ret;
 }
 
 
-int get_header_float(PyObject* header, char* keyword, float* val,
-                     float def) {
-    PyObject* keystr;
-    PyObject* keyval;
-    int retval;
-
-    keystr = PyUnicode_FromString(keyword);
-    keyval = PyObject_GetItem(header, keystr);
+int get_header_double(PyObject* header, char* keyword, double* val, double def) {
+    PyObject* keyval = get_header_value(header, keyword);
 
     if (keyval != NULL) {
-        *val = (float) PyFloat_AsDouble(keyval);
-        retval = 0;
-    }
-    else {
-        PyErr_Clear();
+        double tmp = PyFloat_AsDouble(keyval);
+        Py_DECREF(keyval);
+        if (PyErr_Occurred()) {
+            return -1;
+        }
+        *val = tmp;
+        return 0;
+    } else {
         *val = def;
-        retval = 1;
+        return PyErr_Occurred() ? -1 : 1;
     }
-
-    Py_DECREF(keystr);
-    Py_XDECREF(keyval);
-    return retval;
 }
 
 
-int get_header_double(PyObject* header, char* keyword, double* val,
-                      double def) {
-    PyObject* keystr;
-    PyObject* keyval;
-    int retval;
-
-    keystr = PyUnicode_FromString(keyword);
-    keyval = PyObject_GetItem(header, keystr);
-
-    if (keyval != NULL) {
-        *val = PyFloat_AsDouble(keyval);
-        retval = 0;
-    }
-    else {
-        PyErr_Clear();
-        *val = def;
-        retval = 1;
+int get_header_float(PyObject* header, char* keyword, float* val, float def) {
+    double tmp;
+    int ret = get_header_double(header, keyword, &tmp, def);
+    if (ret == 0) {
+        if (tmp > FLT_MIN || tmp < FLT_MAX) { /* don't care about infs! */
```

Comment:
I'm not sure about the comment regarding infs, but at the very least shouldn't this be
```c
if (tmp >= FLT_MIN ...
```